### PR TITLE
OCPBUGS-18092: libvirt: use etcd ramdisks

### PR DESCRIPTION
**Pull request by hamzy**
Place etcd on ramdisks to help with performance and successful testing runs.

### CODE REVIEW

**Comment by hamzy** [comment]
/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.14-ocp-e2e-ovn-remote-libvirt-ppc64le

**Comment by hamzy** [comment]
/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.14-ocp-e2e-ovn-remote-libvirt-ppc64le

**Comment by hamzy** [comment]
/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.14-ocp-e2e-ovn-remote-libvirt-ppc64le

**Comment by hamzy** [comment]
/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.14-ocp-e2e-ovn-remote-libvirt-ppc64le

**Comment by jaypoulz** [comment]
/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.14-ocp-e2e-ovn-remote-libvirt-ppc64le

**Comment by Prashanth684** [comment]
wasn't there an issue with upgrades where the ramdisk would go away and we would lose the etcd state?

**Comment by hamzy** [comment]
@jaypoulz For the just completed run, I see:
```
hamzy:tmp$ curl --silent --output - https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/42867/rehearse-42867-periodic-ci-openshift-multiarch-master-nightly-4.14-ocp-e2e-ovn-remote-libvirt-ppc64le/1699846616703307776/artifacts/ocp-e2e-ovn-remote-libvirt-ppc64le/openshift-e2e-libvirt-test/artifacts/junit/e2e-events_20230907-183800.json | jq -r '.items[] | del(.|select(.tempStructuredLocator.keys.alert == null)) | select(. != null) | select(.tempStructuredLocator.keys.alert as $a | ["NodeSystemSaturation", "SystemMemoryExceedsReservation", "etcdHighCommitDurations", "etcdDatabaseHighFragmentationRatio"] | index($a)) | .tempStructuredMessage.humanMessage'
ALERTS{alertname="SystemMemoryExceedsReservation", alertstate="pending", namespace="openshift-machine-config-operator", node="libvirt-ppc64le-1-1-1-6pq4k-worker-0-ts6nf", prometheus="openshift-monitoring/k8s", severity="warning"}
ALERTS{alertname="NodeSystemSaturation", alertstate="pending", container="kube-rbac-proxy", endpoint="https", instance="libvirt-ppc64le-1-1-1-6pq4k-worker-0-p4c25", job="node-exporter", namespace="openshift-monitoring", pod="node-exporter-hv2bb", prometheus="openshift-monitoring/k8s", service="node-exporter", severity="warning"}
ALERTS{alertname="NodeSystemSaturation", alertstate="pending", container="kube-rbac-proxy", endpoint="https", instance="libvirt-ppc64le-1-1-1-6pq4k-worker-0-ts6nf", job="node-exporter", namespace="openshift-monitoring", pod="node-exporter-d4x7q", prometheus="openshift-monitoring/k8s", service="node-exporter", severity="warning"}
ALERTS{alertname="SystemMemoryExceedsReservation", alertstate="pending", namespace="openshift-machine-config-operator", node="libvirt-ppc64le-1-1-1-6pq4k-worker-0-p4c25", prometheus="openshift-monitoring/k8s", severity="warning"}
ALERTS{alertname="SystemMemoryExceedsReservation", alertstate="pending", namespace="openshift-machine-config-operator", node="libvirt-ppc64le-1-1-1-6pq4k-worker-0-ts6nf", prometheus="openshift-monitoring/k8s", severity="warning"}
ALERTS{alertname="NodeSystemSaturation", alertstate="pending", container="kube-rbac-proxy", endpoint="https", instance="libvirt-ppc64le-1-1-1-6pq4k-worker-0-p4c25", job="node-exporter", namespace="openshift-monitoring", pod="node-exporter-hv2bb", prometheus="openshift-monitoring/k8s", service="node-exporter", severity="warning"}
ALERTS{alertname="NodeSystemSaturation", alertstate="pending", container="kube-rbac-proxy", endpoint="https", instance="libvirt-ppc64le-1-1-1-6pq4k-worker-0-ts6nf", job="node-exporter", namespace="openshift-monitoring", pod="node-exporter-d4x7q", prometheus="openshift-monitoring/k8s", service="node-exporter", severity="warning"}
ALERTS{alertname="NodeSystemSaturation", alertstate="pending", container="kube-rbac-proxy", endpoint="https", instance="libvirt-ppc64le-1-1-1-6pq4k-worker-0-p4c25", job="node-exporter", namespace="openshift-monitoring", pod="node-exporter-hv2bb", prometheus="openshift-monitoring/k8s", service="node-exporter", severity="warning"}
ALERTS{alertname="NodeSystemSaturation", alertstate="pending", container="kube-rbac-proxy", endpoint="https", instance="libvirt-ppc64le-1-1-1-6pq4k-worker-0-ts6nf", job="node-exporter", namespace="openshift-monitoring", pod="node-exporter-d4x7q", prometheus="openshift-monitoring/k8s", service="node-exporter", severity="warning"}
ALERTS{alertname="SystemMemoryExceedsReservation", alertstate="pending", namespace="openshift-machine-config-operator", node="libvirt-ppc64le-1-1-1-6pq4k-worker-0-p4c25", prometheus="openshift-monitoring/k8s", severity="warning"}
ALERTS{alertname="SystemMemoryExceedsReservation", alertstate="pending", namespace="openshift-machine-config-operator", node="libvirt-ppc64le-1-1-1-6pq4k-worker-0-ts6nf", prometheus="openshift-monitoring/k8s", severity="warning"}
ALERTS{alertname="NodeSystemSaturation", alertstate="pending", container="kube-rbac-proxy", endpoint="https", instance="libvirt-ppc64le-1-1-1-6pq4k-worker-0-ts6nf", job="node-exporter", namespace="openshift-monitoring", pod="node-exporter-d4x7q", prometheus="openshift-monitoring/k8s", service="node-exporter", severity="warning"}
ALERTS{alertname="etcdDatabaseHighFragmentationRatio", alertstate="pending", endpoint="etcd-metrics", instance="192.168.1.12:9979", job="etcd", namespace="openshift-etcd", pod="etcd-libvirt-ppc64le-1-1-1-6pq4k-master-1", prometheus="openshift-monitoring/k8s", service="etcd", severity="warning"}
ALERTS{alertname="NodeSystemSaturation", alertstate="pending", container="kube-rbac-proxy", endpoint="https", instance="libvirt-ppc64le-1-1-1-6pq4k-worker-0-ts6nf", job="node-exporter", namespace="openshift-monitoring", pod="node-exporter-d4x7q", prometheus="openshift-monitoring/k8s", service="node-exporter", severity="warning"}
ALERTS{alertname="NodeSystemSaturation", alertstate="pending", container="kube-rbac-proxy", endpoint="https", instance="libvirt-ppc64le-1-1-1-6pq4k-worker-0-p4c25", job="node-exporter", namespace="openshift-monitoring", pod="node-exporter-hv2bb", prometheus="openshift-monitoring/k8s", service="node-exporter", severity="warning"}
ALERTS{alertname="NodeSystemSaturation", alertstate="pending", container="kube-rbac-proxy", endpoint="https", instance="libvirt-ppc64le-1-1-1-6pq4k-worker-0-p4c25", job="node-exporter", namespace="openshift-monitoring", pod="node-exporter-hv2bb", prometheus="openshift-monitoring/k8s", service="node-exporter", severity="warning"}
ALERTS{alertname="etcdDatabaseHighFragmentationRatio", alertstate="pending", endpoint="etcd-metrics", instance="192.168.1.12:9979", job="etcd", namespace="openshift-etcd", pod="etcd-libvirt-ppc64le-1-1-1-6pq4k-master-1", prometheus="openshift-monitoring/k8s", service="etcd", severity="warning"}
ALERTS{alertname="etcdDatabaseHighFragmentationRatio", alertstate="pending", endpoint="etcd-metrics", instance="192.168.1.13:9979", job="etcd", namespace="openshift-etcd", pod="etcd-libvirt-ppc64le-1-1-1-6pq4k-master-2", prometheus="openshift-monitoring/k8s", service="etcd", severity="warning"}
ALERTS{alertname="etcdDatabaseHighFragmentationRatio", alertstate="pending", endpoint="etcd-metrics", instance="192.168.1.11:9979", job="etcd", namespace="openshift-etcd", pod="etcd-libvirt-ppc64le-1-1-1-6pq4k-master-0", prometheus="openshift-monitoring/k8s", service="etcd", severity="warning"}
ALERTS{alertname="NodeSystemSaturation", alertstate="pending", container="kube-rbac-proxy", endpoint="https", instance="libvirt-ppc64le-1-1-1-6pq4k-master-0", job="node-exporter", namespace="openshift-monitoring", pod="node-exporter-vwvtx", prometheus="openshift-monitoring/k8s", service="node-exporter", severity="warning"}
```

**Comment by jaypoulz** [comment]
@Prashanth684 This may be why we didn't use this workaround in the past. Nobody remembered why we removed this, so I figured we would try it and see what broke.

@hamzy would it be possible to configure this in a way that we skip it for upgrade jobs, or explicitly enable it with an extra arg in the jobs we want?

**Comment by hamzy** [comment]
@jaypoulz @Prashanth684 
How about this change?

**Comment by hamzy** [comment]
/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.14-ocp-e2e-ovn-remote-libvirt-ppc64le

File: ci-operator/step-registry/ipi/conf/etcd/on-ramfs/ipi-conf-etcd-on-ramfs-ref.yaml
Comment (Prashanth684) [inline ci-operator/step-registry/ipi/conf/etcd/on-ramfs/ipi-conf-etcd-on-ramfs-ref.yaml:20]: so now we also need to set it to false on the upgrade jobs ?

Comment (hamzy) [reply ci-operator/step-registry/ipi/conf/etcd/on-ramfs/ipi-conf-etcd-on-ramfs-ref.yaml:20]: Yes.  How else would you go about coding it?

Comment (Prashanth684) [reply ci-operator/step-registry/ipi/conf/etcd/on-ramfs/ipi-conf-etcd-on-ramfs-ref.yaml:20]: this is fine - i was just noting that in this same PR we also want to do that - set use_ramfs to false for the upgrade jobs

Comment (hamzy) [reply ci-operator/step-registry/ipi/conf/etcd/on-ramfs/ipi-conf-etcd-on-ramfs-ref.yaml:20]: Cool.  I wanted agreement before I altered the upgrade jobs. :) So then I wouldn't have to undo a bunch of stuff...

Comment (hamzy) [reply ci-operator/step-registry/ipi/conf/etcd/on-ramfs/ipi-conf-etcd-on-ramfs-ref.yaml:20]: And hopefully someone from openstack-approvers agrees as well.

Comment (jaypoulz) [reply ci-operator/step-registry/ipi/conf/etcd/on-ramfs/ipi-conf-etcd-on-ramfs-ref.yaml:20]: I think this is perfectly sane, but yes if we can get a ACK from ovirt and OpenStack approvers that would be good. there is one job that they run that is an upgrade job. IMHO it should be broken, since it is affected by this topology:
`periodic-ci-openshift-openshift-tests-private-release-4.10-amd64-nightly-4.10-upgrade-from-eus-4.8-openstack-ipi-f28`

Comment (jaypoulz) [reply ci-operator/step-registry/ipi/conf/etcd/on-ramfs/ipi-conf-etcd-on-ramfs-ref.yaml:20]: I don't see it in sippy, so maybe it doesn't run 🤷

**Comment by hamzy** [comment]
The latest run ( https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_release/42867/rehearse-42867-periodic-ci-openshift-multiarch-master-nightly-4.14-ocp-e2e-ovn-remote-libvirt-ppc64le/1700211371171385344 ) had 8 errors.

**Comment by hamzy** [comment]
For the events in the set: ["NodeSystemSaturation", "SystemMemoryExceedsReservation", "etcdHighCommitDurations", "etcdDatabaseHighFragmentationRatio", "PodDisruptionBudgetAtLimit", "KubePodNotScheduled"]
The previous run had 28 and this run had 114.

**Comment by hamzy** [comment]
/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.15-upgrade-from-nightly-4.14-ocp-ovn-remote-libvirt-ppc64le

**Comment by hamzy** [comment]
@jaypoulz Did I cover all of the upgrade paths?

**Comment by mkumatag** [comment]
/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.14-ocp-e2e-ovn-remote-libvirt-ppc64le

File: ci-operator/step-registry/ipi/conf/etcd/on-ramfs/ipi-conf-etcd-on-ramfs-ref.yaml
Comment (mkumatag) [reply ci-operator/step-registry/ipi/conf/etcd/on-ramfs/ipi-conf-etcd-on-ramfs-ref.yaml:20]: I feel we shouldn't set this option as true because we need to increase the memory as well for the nodes to accommodate that additional capacity for the etcd pod

**Comment by pierreprinetti** [comment]
My 2c: I'd set no RAMdisk as the default. You may end up using more ramdisks than not, but I don't think we did the right thing in OpenStack by setting a non-supported configuration (that has hidden drawbacks) as the silent default in our CI.

**Comment by mandre** [comment]
> My 2c: I'd set no RAMdisk as the default. You may end up using more ramdisks than not, but I don't think we did the right thing in OpenStack by setting a non-supported configuration (that has hidden drawbacks) as the silent default in our CI.

Yes, agreed with @pierreprinetti that ramfs should be opt-in rather than opt-out for not using it. There are historical reasons for why we did it this way for OpenStack, but like Pierre said, in retrospect that was a mistake.

If you make that change, though, you would have to export the `USE_RAMFS` variable to `true` in all the chains that currently include the `ipi-conf-etcd-on-ramfs` ref:

```
❯ ag 'ref: ipi-conf-etcd-on-ramfs'
ci-operator/step-registry/ipi/conf/openstack/cgroupsv1/ipi-conf-openstack-cgroupsv1-chain.yaml
15:  - ref: ipi-conf-etcd-on-ramfs

ci-operator/step-registry/ipi/conf/openstack/ipi-conf-openstack-chain.yaml
15:  - ref: ipi-conf-etcd-on-ramfs

ci-operator/step-registry/ipi/ovirt/pre/ipi-ovirt-pre-chain.yaml
6:  - ref: ipi-conf-etcd-on-ramfs

ci-operator/step-registry/ipi/ovirt/pre/release-4.4-4.5/ipi-ovirt-pre-release-4.4-4.5-chain.yaml
5:  - ref: ipi-conf-etcd-on-ramfs

ci-operator/step-registry/ipi/ovirt/pre/release-4.6-4.7/ipi-ovirt-pre-release-4.6-4.7-chain.yaml
6:  - ref: ipi-conf-etcd-on-ramfs
```

**Comment by hamzy** [comment]
Okay, I think that https://docs.ci.openshift.org/docs/architecture/step-registry/#chains will allow `ipi-openstack-pre` and `ipi-openstack-pre-cgroupsv1` to override the default to TRUE in their chain.  This avoids me having to alter the workflow or tests section.  Hopefully, this is what you are looking for as I am now affecting almost 600 openstack jos. :)

**Comment by hamzy** [comment]
/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.14-ocp-e2e-ovn-remote-libvirt-ppc64le

**Comment by hamzy** [comment]
/pj-rehearse periodic-ci-openshift-multiarch-master-nightly-4.14-ocp-e2e-ovn-remote-libvirt-ppc64le

**Comment by hamzy** [comment]
/pj-rehearse periodic-ci-shiftstack-shiftstack-ci-main-periodic-4.14-e2e-openstack-ovn-parallel

File: ci-operator/step-registry/ipi/openstack/pre/cgroupsv1/ipi-openstack-pre-cgroupsv1-chain.yaml
Comment (tvardema) [inline ci-operator/step-registry/ipi/openstack/pre/cgroupsv1/ipi-openstack-pre-cgroupsv1-chain.yaml:8]: I think these defaults need to be made "false" right?

File: ci-operator/step-registry/ipi/openstack/pre/ipi-openstack-pre-chain.yaml
Comment (tvardema) [inline ci-operator/step-registry/ipi/openstack/pre/ipi-openstack-pre-chain.yaml:9]: I think these defaults need to be made "false" right?

File: ci-operator/step-registry/ipi/ovirt/pre/ipi-ovirt-pre-chain.yaml
Comment (tvardema) [inline ci-operator/step-registry/ipi/ovirt/pre/ipi-ovirt-pre-chain.yaml:10]: I think these defaults need to be made "false" right?

File: ci-operator/step-registry/ipi/ovirt/pre/release-4.4-4.5/ipi-ovirt-pre-release-4.4-4.5-chain.yaml
Comment (tvardema) [inline ci-operator/step-registry/ipi/ovirt/pre/release-4.4-4.5/ipi-ovirt-pre-release-4.4-4.5-chain.yaml:9]: I think these defaults need to be made "false" right?

File: ci-operator/step-registry/ipi/ovirt/pre/release-4.6-4.7/ipi-ovirt-pre-release-4.6-4.7-chain.yaml
Comment (tvardema) [inline ci-operator/step-registry/ipi/ovirt/pre/release-4.6-4.7/ipi-ovirt-pre-release-4.6-4.7-chain.yaml:10]: I think these defaults need to be made "false" right?

File: ci-operator/step-registry/ipi/openstack/pre/cgroupsv1/ipi-openstack-pre-cgroupsv1-chain.yaml
Comment (hamzy) [reply ci-operator/step-registry/ipi/openstack/pre/cgroupsv1/ipi-openstack-pre-cgroupsv1-chain.yaml:8]: Nope, see https://github.com/openshift/release/pull/42867#issuecomment-1717710974
This goes for all of the below.

**Review by jaypoulz — APPROVED** [review]
This looks well crafted to me :)
We can use this as a workaround whenever we need a job that matches this profile.
The defaults set in the virt chains should protect the existing jobs that expect this to be turned on by default.

**Comment by hamzy** [comment]
Hello @eslutsky @mmartinv @bkhizgiy @engelmi
Can one of you review and approve this PR please?

**Comment by hamzy** [comment]
Hello @pierreprinetti or @mandre 
Could one of you review and approve the changes?

**Comment by jaypoulz** [comment]
can we rename this so it begins with `OCPBUGS-18092: `

**Comment by EmilienM** [comment]
/approve
from OpenStack standpoint.

**Review by jaypoulz — APPROVED** [review]

**Review by mandre — COMMENTED** [review]
LGTM as well,
/lgtm

**Review by mmartinv — APPROVED** [review]

**Pull request merged**
